File: NoTomorrow/Services/AuthStore.swift
import Foundation
import Observation
import Security

/// Session and the optional Anthropic key, both in the Keychain (never iCloud-synchronised).
/// One instance per process (`shared`): the sign-in sheet, Settings, the Bro tab and the AI flow all observe it.
@Observable
@MainActor
final class AuthStore {
    static let shared = AuthStore()

    private(set) var session: Session?
    /// Masked Anthropic key for display, e.g. "sk-ant-…1234". `nil` when no key is stored.
    private(set) var maskedAnthropicKey: String?

    var isSignedIn: Bool { session != nil }
    var hasAnthropicKey: Bool { maskedAnthropicKey != nil }

    /// `true` when the real backend is in use and there is no session: Bro pairing and AI estimates
    /// show their "Sign in" state instead of calling the network. The demo backend never needs an account.
    var needsSignIn: Bool { !AppConfig.shared.useMockBackend && session == nil }

    init() {
        session = KeychainHelper.readSession()
        maskedAnthropicKey = KeychainHelper.readAnthropicKey().map(Self.mask)
    }

    func save(_ session: Session) {
        KeychainHelper.writeSession(session)
        self.session = session
    }

    func clear() {
        KeychainHelper.delete(account: KeychainHelper.Account.session)
        session = nil
    }

    /// Re-reads the Keychain after the transport rotated (or dropped) the session off the main actor.
    func reload() {
        session = KeychainHelper.readSession()
    }

    // MARK: Anthropic key (BYOK)

    /// The raw key. Read on demand for the request header; never held in a view.
    var anthropicKey: String? {
        get { KeychainHelper.readAnthropicKey() }
        set {
            if let key = newValue?.trimmingCharacters(in: .whitespacesAndNewlines), !key.isEmpty {
                KeychainHelper.writeAnthropicKey(key)
                maskedAnthropicKey = Self.mask(key)
            } else {
                removeAnthropicKey()
            }
        }
    }

    func removeAnthropicKey() {
        KeychainHelper.delete(account: KeychainHelper.Account.anthropicKey)
        maskedAnthropicKey = nil
    }

    /// "sk-ant-api03-…7Yq2" → "sk-ant-…7Yq2"
    static func mask(_ key: String) -> String {
        let suffix = String(key.suffix(4))
        let prefix = key.hasPrefix("sk-ant-") ? "sk-ant-" : String(key.prefix(min(3, max(0, key.count - 4))))
        return "\(prefix)…\(suffix)"
    }
}

// MARK: - Keychain

/// Minimal generic-password wrapper. Items are device-only and never synchronisable.
enum KeychainHelper {
    static let service = "app.notomorrow"

    enum Account {
        static let session = "session"
        static let anthropicKey = "anthropic-key"
    }

    // Session: available once the device has been unlocked after boot (background refreshes, push handling).
    static func readSession() -> Session? {
        guard let data = read(account: Account.session) else { return nil }
        return try? JSONDecoder().decode(Session.self, from: data)
    }

    static func writeSession(_ session: Session) {
        guard let data = try? JSONEncoder().encode(session) else { return }
        write(data, account: Account.session, accessible: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly)
    }

    // Anthropic key: only while unlocked; the user is always in the foreground when it is used.
    static func readAnthropicKey() -> String? {
        read(account: Account.anthropicKey).flatMap { String(data: $0, encoding: .utf8) }
    }

    static func writeAnthropicKey(_ key: String) {
        write(Data(key.utf8), account: Account.anthropicKey, accessible: kSecAttrAccessibleWhenUnlockedThisDeviceOnly)
    }

    // MARK: Primitives

    static func read(account: String) -> Data? {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
            kSecAttrSynchronizable: false,
            kSecReturnData: true,
            kSecMatchLimit: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess else { return nil }
        return result as? Data
    }

    @discardableResult
    static func write(_ data: Data, account: String, accessible: CFString) -> Bool {
        let base: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
            kSecAttrSynchronizable: false,
        ]
        let attributes: [CFString: Any] = [kSecValueData: data, kSecAttrAccessible: accessible]
        let update = SecItemUpdate(base as CFDictionary, attributes as CFDictionary)
        if update == errSecSuccess { return true }
        if update != errSecItemNotFound {
            // Accessibility can't change through an update; replace the item.
            SecItemDelete(base as CFDictionary)
        }
        let add = base.merging(attributes) { _, new in new }
        return SecItemAdd(add as CFDictionary, nil) == errSecSuccess
    }

    @discardableResult
    static func delete(account: String) -> Bool {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
            kSecAttrSynchronizable: false,
        ]
        let status = SecItemDelete(query as CFDictionary)
        return status == errSecSuccess || status == errSecItemNotFound
    }
}

File: android/app/src/main/java/app/notomorrow/service/AuthStore.kt
package app.notomorrow.service

import app.notomorrow.data.prefs.SecureStore
import app.notomorrow.net.SessionStore
import app.notomorrow.net.dto.Session
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

/**
 * Session and the optional BYOK keys (Anthropic, Gemini) — the port of `AuthStore` in
 * `NoTomorrow/Services/AuthStore.swift`, with `SecureStore` (Keystore + Tink + DataStore)
 * standing in for the Keychain.
 *
 * One instance per process: the sign-in sheet, Settings, the Bro tab and the AI flow all observe it.
 * It also **is** the [SessionStore] the transport reads on the hot path of every authenticated call
 * — the analogue of `AppConfig.keychainSessionStorage`, whose write closure nudges the observable
 * store. Because `SecureStore` is `suspend` and [SessionStore.session] must not be, the session is
 * mirrored into a `StateFlow` and written back asynchronously.
 *
 * [load] must run once at container construction, before the first authenticated call.
 */
class AuthStore(
    private val secureStore: SecureStore,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) : SessionStore {

    private val _session = MutableStateFlow<Session?>(null)
    val sessionFlow: StateFlow<Session?> = _session.asStateFlow()

    val isSignedIn: StateFlow<Boolean> =
        _session.map { it != null }.stateIn(scope, SharingStarted.Eagerly, false)

    /** Masked Anthropic key for display, e.g. `"sk-ant-…7Yq2"`. `null` when no key is stored. */
    private val _maskedAnthropicKey = MutableStateFlow<String?>(null)
    val maskedAnthropicKey: StateFlow<String?> = _maskedAnthropicKey.asStateFlow()

    val hasAnthropicKey: StateFlow<Boolean> =
        _maskedAnthropicKey.map { it != null }.stateIn(scope, SharingStarted.Eagerly, false)

    /** Masked Gemini key for display, e.g. `"AIz…7Yq2"`. `null` when no key is stored. */
    private val _maskedGeminiKey = MutableStateFlow<String?>(null)
    val maskedGeminiKey: StateFlow<String?> = _maskedGeminiKey.asStateFlow()

    val hasGeminiKey: StateFlow<Boolean> =
        _maskedGeminiKey.map { it != null }.stateIn(scope, SharingStarted.Eagerly, false)

    /** Fed by `AppConfig.useMockBackend` through [bindMockBackend]; the demo backend needs no account. */
    private val mockBackend = MutableStateFlow(false)

    /**
     * `true` when the real backend is in use and there is no session: Bro pairing and AI estimates
     * show their "Sign in" state instead of calling the network.
     */
    val needsSignIn: StateFlow<Boolean> =
        combine(_session, mockBackend) { session, mock -> !mock && session == null }
            .stateIn(scope, SharingStarted.Eagerly, true)

    // MARK: - Lifecycle

    /** The `init` half of the Swift store: read every record out of the secure store. */
    suspend fun load() {
        _session.value = secureStore.session()
        _maskedAnthropicKey.value = secureStore.anthropicKey()?.let(::mask)
        _maskedGeminiKey.value = secureStore.geminiKey()?.let(::mask)
    }

    /** Wired in `AppContainer`: `authStore.bindMockBackend(appConfig.useMockBackend)`. */
    fun bindMockBackend(flow: Flow<Boolean>) {
        scope.launch { flow.collect { mockBackend.value = it } }
    }

    /** Re-reads the store after the transport rotated (or dropped) the session off this class. */
    suspend fun reload() {
        _session.value = secureStore.session()
    }

    // MARK: - SessionStore (synchronous, any thread)

    override fun session(): Session? = _session.value

    override fun save(session: Session) {
        _session.value = session
        scope.launch { secureStore.setSession(session) }
    }

    override fun clear() {
        _session.value = null
        scope.launch { secureStore.clearSession() }
    }

    /** The sign-in path: the session must be durable before the next call goes out. */
    suspend fun saveAwait(session: Session) {
        secureStore.setSession(session)
        _session.value = session
    }

    /** Sign-out / delete-account: wait for the record to be gone. */
    suspend fun clearAwait() {
        secureStore.clearSession()
        _session.value = null
    }

    // MARK: - Anthropic key (BYOK)

    /** The raw key. Read on demand for the request header; never held in a view. */
    suspend fun anthropicKey(): String? = secureStore.anthropicKey()

    /** A blank key removes the record, exactly as the Swift setter does. */
    suspend fun setAnthropicKey(key: String?) {
        val trimmed = key?.trim()
        if (trimmed.isNullOrEmpty()) {
            removeAnthropicKey()
            return
        }
        secureStore.setAnthropicKey(trimmed)
        _maskedAnthropicKey.value = mask(trimmed)
    }

    suspend fun removeAnthropicKey() {
        secureStore.clearAnthropicKey()
        _maskedAnthropicKey.value = null
    }

    // MARK: - Gemini key (BYOK)

    /** The raw key. Read on demand for the request header; never held in a view. */
    suspend fun geminiKey(): String? = secureStore.geminiKey()

    /** A blank key removes the record, exactly as the Swift setter does. */
    suspend fun setGeminiKey(key: String?) {
        val trimmed = key?.trim()
        if (trimmed.isNullOrEmpty()) {
            removeGeminiKey()
            return
        }
        secureStore.setGeminiKey(trimmed)
        _maskedGeminiKey.value = mask(trimmed)
    }

    suspend fun removeGeminiKey() {
        secureStore.clearGeminiKey()
        _maskedGeminiKey.value = null
    }

    companion object {
        /** `"sk-ant-api03-…7Yq2"` → `"sk-ant-…7Yq2"`. */
        fun mask(key: String): String {
            val suffix = key.takeLast(4)
            val prefix = if (key.startsWith("sk-ant-")) {
                "sk-ant-"
            } else {
                key.take(minOf(3, maxOf(0, key.length - 4)))
            }
            return "$prefix…$suffix"
        }
    }
}
